### PR TITLE
feat(parser): parse modifiers with `parse_modifiers`

### LINKED ISSUE #3929: parser: finish parsing of `Modifiers`
https://github.com/oxc-project/oxc/blob/main/crates/oxc_parser/src/modifiers.rs

### PR DESCRIPTION
**Pull request by DonIsaac**
Closes #3929

### CODE REVIEW

File: crates/oxc_parser/src/modifiers.rs [hunk 1]
```diff
@@ -102,39 +124,125 @@ pub struct Modifier {
...
+    }
+}
+
+impl GetSpan for Modifiers<'_> {
+    fn span(&self) -> Span {
+        let Some(modifiers) = &self.modifiers else { return SPAN };
+        // SAFETY: One of Modifier's invariants is that Some(modifiers) always
+        // contains a non-empty Vec; otherwise it must be `None`.
+        unsafe {
+            use assert_unchecked::assert_unchecked;
```
Comment (DonIsaac) [inline crates/oxc_parser/src/modifiers.rs:242]: @Boshen is it OK to use `unsafe` here? Or should I use safe code?

File: crates/oxc_parser/src/modifiers.rs [hunk 2]
```diff
@@ -178,20 +286,52 @@ impl ModifierKind {
...
         let mut modifiers = self.ast.new_vec();
         while self.at_modifier() {
             let span = self.start_span();
             let modifier_flag = self.cur_kind().into();
-            flags.set(modifier_flag, true);
             let kind = self.cur_kind();
             self.bump_any();
-            modifiers.push(Self::modifier(kind, self.end_span(span)));
+            let modifier = Self::modifier(kind, self.end_span(span));
+            self.verify_modifier(flags, &modifier);
```
Comment (DonIsaac) [inline crates/oxc_parser/src/modifiers.rs:330]: I'd rather do this in the checker, but we don't persist modifiers in the AST and so we lose spans

File: crates/oxc_parser/src/modifiers.rs (hunk 1 above)
Comment (Boshen) [reply crates/oxc_parser/src/modifiers.rs:208]: `debug_assert` should be fine here?

File: crates/oxc_parser/src/modifiers.rs [hunk 3]
```diff
@@ -178,20 +252,52 @@ impl ModifierKind {
         }
     }
 }
+impl From<Kind> for ModifierKind {
+    // NOTE (@DonIsaac) should we be using TryFrom here? or does unreachable!()
```
Comment (Boshen) [inline crates/oxc_parser/src/modifiers.rs:256]: `unreachable` is too unsafe here, better to return a None, and return unexpected token on the call site.

Need to be extra safe for a parser.

File: crates/oxc_parser/src/modifiers.rs [hunk 4]
```diff
@@ -178,20 +252,52 @@ impl ModifierKind {
...
+
+impl std::fmt::Display for ModifierKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.as_str())
+    }
+}
 
 impl<'a> ParserImpl<'a> {
-    pub(crate) fn eat_modifiers_before_declaration(&mut self) -> (ModifierFlags, Modifiers<'a>) {
+    pub(crate) fn eat_modifiers_before_declaration(&mut self) -> Modifiers<'a> {
```
Comment (Boshen) [inline crates/oxc_parser/src/modifiers.rs:287]: I'd eventually align all the method names with tsc's parser.ts as well so we know where to look when we try to debug parsing.

File: crates/oxc_parser/src/modifiers.rs [hunk 5]
```diff
@@ -231,55 +337,40 @@ impl<'a> ParserImpl<'a> {
...
+            if modifier.is_static() {
                 has_seen_static_modifier = true;
             }
+            self.check_for_duplicate_modifiers(modifier_flags, &modifier);
+            modifier_flags.set(modifier.kind.into(), true);
             modifiers.push(modifier);
+            has_leading_modifier = true;
         }
 
         // parse trailing decorators, but only if we parsed any leading modifiers
```
Comment (Boshen) [inline crates/oxc_parser/src/modifiers.rs:376]: I wonder where does trailing decorators occur lol

File: crates/oxc_parser/src/modifiers.rs [hunk 6]
```diff
@@ -289,47 +380,63 @@ impl<'a> ParserImpl<'a> {
...
+        {
             return None;
         }
-        // }
         Some(Self::modifier(kind, self.end_span(span)))
     }
 
+    fn next_token_is_open_brace(&mut self) -> bool {
+        self.bump_any();
+        self.cur_kind() == Kind::LCurly
```
Comment (Boshen) [inline crates/oxc_parser/src/modifiers.rs:437]: There's a shorthand `self.at(Kind::LCurly)`

**Review by Boshen — CHANGES_REQUESTED** [review]

File: tasks/transform_conformance/babel.snap.md [hunk 7]
```diff
@@ -445,7 +445,9 @@ Passed: 475/927
 * opts/optimizeConstEnums/input.ts
 * opts/rewriteImportExtensions/input.ts
 
-# babel-plugin-transform-typescript (131/151)
+# babel-plugin-transform-typescript (129/151)
+* class/accessor-allowDeclareFields-false/input.ts
```
[the remark this answers is not in the document]
Comment (DonIsaac) [reply tasks/transform_conformance/babel.snap.md:449]: I'm not sure, those cases are using `private` on private class elements.

```ts
abstract class Foo {
  accessor prop: number = 1;
  static accessor prop2: number = 1;
  accessor #prop3: number = 1;
  accessor [prop4]: number = 1;
  private accessor prop5: number = 1;
  abstract accessor prop6: number;
  private accessor #p: any;

  accessor a!: any;
  accessor aa!: any;
  abstract accessor #s;
  abstract accessor f = 1;
  readonly accessor g;
}
```

File: crates/oxc_parser/src/modifiers.rs (hunk 4 above)
Comment (DonIsaac) [reply crates/oxc_parser/src/modifiers.rs:287]: I couldn't figure out what exactly was going on with this method so I left it as is, simply changing the return signature...

File: crates/oxc_parser/src/modifiers.rs (hunk 5 above)
Comment (DonIsaac) [reply crates/oxc_parser/src/modifiers.rs:376]: I think this is b/c of the syntax discrepancy between ts experimental decorators and
the ecmascript proposal?

File: crates/oxc_parser/src/js/class.rs [hunk 8]
```diff
@@ -25,7 +25,8 @@ impl<'a> ParserImpl<'a> {
         stmt_ctx: StatementContext,
         start_span: Span,
     ) -> Result<Statement<'a>> {
-        let decl = self.parse_class_declaration(start_span, &Modifiers::empty())?;
+        let modifiers = self.parse_modifiers(true, false, true);
```
[the remark this answers is not in the document]
Comment (DonIsaac) [reply crates/oxc_parser/src/js/class.rs]: fixed

File: crates/oxc_parser/src/modifiers.rs (hunk 3 above)
Comment (DonIsaac) [reply crates/oxc_parser/src/modifiers.rs:257]: Fixed

File: crates/oxc_parser/src/modifiers.rs (hunk 4 above)
Comment (Boshen) [reply crates/oxc_parser/src/modifiers.rs]: > I couldn't figure out what exactly was going on with this method

Me neither, I didn't write it. We can get rid of it and port from tsc instead.

File: crates/oxc_parser/src/js/list.rs [hunk 9]
```diff
@@ -265,6 +265,7 @@ impl<'a> SeparatedList<'a> for FormalParameterList<'a> {
         let span = p.start_span();
         p.eat_decorators()?;
 
+        // TODO: use parse_modifiers
```
Comment (Boshen) [inline crates/oxc_parser/src/js/list.rs:268]: Better to not have a TODO in the parser.

File: crates/oxc_parser/src/modifiers.rs [hunk 10]
```diff
@@ -76,65 +98,117 @@ impl ModifierFlags {
...
+    /// # Invariants
+    /// `flags` must correctly reflect the [`ModifierKind`]s within
+    ///  `modifiers`. E.g., if `modifiers` is empty, then so is `flags``.
+    #[must_use]
+    fn new(modifiers: Vec<'a, Modifier>, flags: ModifierFlags) -> Self {
+        if modifiers.is_empty() {
+            debug_assert!(flags.is_empty());
+            Self::empty()
+        } else {
+            // todo: loop over `modifiers` and check integrity of `flags`
```
Comment (Boshen) [inline crates/oxc_parser/src/modifiers.rs:159]: Better to not have a TODO in the parser.

File: crates/oxc_parser/src/diagnostics.rs [hunk 11]
```diff
@@ -399,7 +401,35 @@ pub fn jsx_element_no_match(span0: Span, span1: Span, name: &str) -> OxcDiagnost
...
+#[cold]
+pub fn modifier_cannot_be_used_here(modifier: &Modifier) -> OxcDiagnostic {
+    OxcDiagnostic::error(format!("'{}' modifier cannot be used here.", modifier.kind))
+        .with_label(modifier.span)
+}
+
+/// TS(1030)
+#[cold]
+pub fn modifier_already_seen(modifier: &Modifier) -> OxcDiagnostic {
+    OxcDiagnostic::error(format!("'{}' modifier already seen.", modifier.kind))
```
[the remark this answers is not in the document]
Comment (DonIsaac) [reply crates/oxc_parser/src/diagnostics.rs]: I'm thinking of adding error codes to `OxcDiagnostic`, what do you think?

File: crates/oxc_parser/src/modifiers.rs (hunk 4 above)
Comment (DonIsaac) [reply crates/oxc_parser/src/modifiers.rs]: Ive spent two hours staring at this code, `parse_modifers` and typescript's implementation. When I try to use `parse_modifiers` where this function is consumed, I get an infinite loop over `Kind::Const`. I can't figure it out. Any insights?

File: crates/oxc_parser/src/js/list.rs (hunk 9 above)
Comment (DonIsaac) [reply crates/oxc_parser/src/js/list.rs]: fixed

**Review by Boshen — APPROVED** [review]

File: crates/oxc_parser/src/modifiers.rs (hunk 4 above)
Comment (Boshen) [reply crates/oxc_parser/src/modifiers.rs]: Let's track this in the issue.

File: crates/oxc_parser/src/diagnostics.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Boshen) [reply crates/oxc_parser/src/diagnostics.rs]: It's unnecessary for us, we don't produce i18n error messages.

**Pull request merged**